<a href="https://colab.research.google.com/github/shourav47/Best_Bowler_Analysis_in_Test_Cricket/blob/main/best_bowling_in_cricket_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###mounting the drive

In [386]:
from google.colab import drive
drive._mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###importing necessary librarires

In [387]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_rows', 300)
sns.set_context('talk')

###reading the dataset

In [388]:
df = pd.read_csv('/content/drive/MyDrive/dataset/best_bowling_dataset.csv')
display(df.head(10))

Player       Span  Mat  Inns  ...  Econ    SR   5  10
0  M Muralitharan (ICC/SL)  1992-2010  133   230  ...  2.47  55.0  67  22
1           SK Warne (AUS)  1992-2007  145   273  ...  2.65  57.4  37  10
2        JM Anderson (ENG)  2003-2022  169   314  ...  2.80  56.8  31   3
3         A Kumble (INDIA)  1990-2008  132   236  ...  2.69  65.9  35   8
4         GD McGrath (AUS)  1993-2007  124   243  ...  2.49  51.9  29   3
5          SCJ Broad (ENG)  2007-2022  151   278  ...  2.93  57.0  19   3
6            CA Walsh (WI)  1984-2001  132   242  ...  2.53  57.8  22   3
7            DW Steyn (SA)  2004-2019   93   171  ...  3.24  42.3  26   5
8      N Kapil Dev (INDIA)  1978-1994  131   227  ...  2.78  63.9  23   2
9        HMRKB Herath (SL)  1999-2018   93   170  ...  2.80  60.0  34   9

[10 rows x 14 columns]

###Column description:<br>
**Player:** Name of the Player<br>
**Span:** Duration of year a player has player<br>
**Match:** Total matches a player has played<br>
**Inns:** total innings a player has played (generally two innings played in a match by a player).<br>
**Balls:** Total balls delivered by a player.<br>
**Runs:** Total runs has given the player in his entire carrier.<br>
**Wkts:** Total wickets taken by a player<br>
**BBI:** Best bowling a player has done in a innings.<br>
**BBM:** Best bowling a player has done in a match.<br>
**Ave:** Average runs a player has given in a innings.<br>
**Ecn:** average run given by a player in a over.<br>
**Sr:** Strike rate. takes a Wicket after giving this run.<br>
**5:** No of matches in which 5 wickets have taken.<br>
**10:** No of matches in which 10 wickets have taken. <br>


In [389]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  80 non-null     object 
 1   Span    80 non-null     object 
 2   Mat     80 non-null     object 
 3   Inns    80 non-null     int64  
 4   Balls   80 non-null     int64  
 5   Runs    80 non-null     int64  
 6   Wkts    80 non-null     int64  
 7   BBI     80 non-null     object 
 8   BBM     80 non-null     object 
 9   Ave     80 non-null     float64
 10  Econ    80 non-null     float64
 11  SR      80 non-null     float64
 12  5       80 non-null     int64  
 13  10      80 non-null     int64  
dtypes: float64(3), int64(6), object(5)
memory usage: 8.9+ KB


In [390]:
print('no of rows:', df.shape[0])
print('no of columns:', df.shape[1])

no of rows: 80
no of columns: 14


The above dataframe has 80 rows has 14 column.<br>
it has 8 neumerical columns & 6 string type column<br>

In [391]:
df.describe()

Inns         Balls          Runs  ...         SR         5         10
count   80.000000     80.000000     80.000000  ...  80.000000  80.00000  80.000000
mean   145.362500  18640.687500   8605.462500  ...  59.122500  16.27500   2.762500
std     51.452139   7245.216438   3100.829672  ...   9.354955   9.65739   3.230565
min     67.000000   9266.000000   5000.000000  ...  41.000000   3.00000   0.000000
25%    108.750000  13627.250000   6379.750000  ...  53.275000   9.00000   1.000000
50%    129.000000  16514.500000   7708.000000  ...  57.200000  14.00000   2.000000
75%    168.500000  21826.750000   9744.500000  ...  63.925000  20.25000   3.250000
max    314.000000  44039.000000  18355.000000  ...  91.900000  67.00000  22.000000

[8 rows x 9 columns]

description:

25% player have played 108 innings or less.<br>
50% player have played 129 innings or less.<br>
75% player have played 168 innings or less.

#renaming the column name

In [392]:
df.columns

Index(['Player', 'Span', 'Mat', 'Inns', 'Balls', 'Runs', 'Wkts', 'BBI', 'BBM',
       'Ave', 'Econ', 'SR', '5', '10'],
      dtype='object')

In [393]:
# columns = ['Player', 'Span', 'Match', 'Innings', 'Total_Balls', 'Total_Runs', 'Wkts_Taken', 'Best_Bowlings_Per_Innings', 'Best_Bowling_Match',
#        'Average', 'Economy', 'StrikeRate', '5_wkts_in_match', '10_wkts_in_match']

df = df.rename(columns = {'Player':'Player', 'Span':'Span', 'Mat':'Match', 'Inns':'Innings', 'Balls':'Total_Balls', 'Runs':'Total_Runs', 'Wkts':'Wickets', 'BBI':'BBI', 'BBM':'BBM','Ave':'Average', 'Econ':'Economy', 'SR':'StrikeRate', '5':'5_wkts_in_match', '10':'10_wkts_in_match'})

In [394]:
# df.columns=[columns]
display(df.head(2))

#We can do the same rename task by using dictonary(rename function)

Player       Span  ... 5_wkts_in_match  10_wkts_in_match
0  M Muralitharan (ICC/SL)  1992-2010  ...              67                22
1           SK Warne (AUS)  1992-2007  ...              37                10

[2 rows x 14 columns]

#Question on Mind
1. Separate only player name
2. create a column with player based on their country
3. Find if the player have played for icc or not.
4. Arrenge the dataset according to most played match

6. Find the player who have given the highest run
7. Find the no of years played by a player.
8. find who have best bowlings per innings
9. plot top 10 player with top strike rate
10. Replace Boolean 'Played_for_ICC' into category
11. Find the top player who have most

####First Question: Separate only player name

In [395]:
# df['Player'] = df['Player'].str.replace('(','')
#printing the values inside a feature
df['Player'].values

array(['M Muralitharan (ICC/SL)', 'SK Warne (AUS)', 'JM Anderson (ENG)',
       'A Kumble (INDIA)', 'GD McGrath (AUS)', 'SCJ Broad (ENG)',
       'CA Walsh (WI)', 'DW Steyn (SA)', 'N Kapil Dev (INDIA)',
       'HMRKB Herath (SL)', 'Sir RJ Hadlee (NZ)', 'R Ashwin (INDIA)',
       'SM Pollock (SA)', 'Harbhajan Singh (INDIA)', 'NM Lyon (AUS)',
       'Wasim Akram (PAK)', 'CEL Ambrose (WI)', 'M Ntini (SA)',
       'IT Botham (ENG)', 'MD Marshall (WI)', 'Waqar Younis (PAK)',
       'Imran Khan (PAK)', 'DL Vettori (ICC/NZ)', 'DK Lillee (AUS)',
       'WPUJC Vaas (SL)', 'AA Donald (SA)', 'TG Southee (NZ)',
       'RGD Willis (ENG)', 'MG Johnson (AUS)', 'I Sharma (INDIA)',
       'Z Khan (INDIA)', 'B Lee (AUS)', 'M Morkel (SA)', 'LR Gibbs (WI)',
       'FS Trueman (ENG)', 'TA Boult (NZ)', 'DL Underwood (ENG)',
       'JH Kallis (ICC/SA)', 'CJ McDermott (AUS)', 'MA Starc (AUS)',
       'BS Bedi (INDIA)', 'Danish Kaneria (PAK)', 'J Garner (WI)',
       'JN Gillespie (AUS)', 'GP Swann (ENG)', 'JB

In [396]:
#splitting the player column
df_player = df['Player'].str.split('(',expand = True)
#renaming the column name
df_player = df_player.rename( columns={0:'Player_name', 1:'Country'})
df_player['Country'] = df_player['Country'].str.replace(')','')
display(df_player.head(5))
print(type(df_player))

Player_name Country
0  M Muralitharan   ICC/SL
1        SK Warne      AUS
2     JM Anderson      ENG
3        A Kumble    INDIA
4      GD McGrath      AUS

<class 'pandas.core.frame.DataFrame'>


In [397]:
#concatation of two dataframe
df = pd.concat([df,df_player], axis= 1)
#droping one redundant column
df = df.drop('Player', axis=1, )
display(df.head(2))

Span Match  Innings  ...  10_wkts_in_match      Player_name  Country
0  1992-2010   133      230  ...                22  M Muralitharan    ICC/SL
1  1992-2007   145      273  ...                10        SK Warne       AUS

[2 rows x 15 columns]

In [398]:
#display(df)

####Ques 2: Checking if the player have played for icc or not

In [399]:
#finding the answer through pandas apply method
def icc(x):
  if 'ICC/' in x:
    return 1
  elif '/ICC' in x:
    return 1
  else:
    return 0
df['Played_for_ICC'] = df['Country'].apply(icc)
#display(df)
  

####Ques 3: create a column with player based on their country

In [400]:
#removing unwanted alphabet from a column
df['Country'] = df['Country'].str.replace('ICC/','')
df['Country'] = df['Country'].str.replace('/ICC','')
df['Match'] = df['Match'].str.replace('*','')
#display(df)

In [401]:
df.columns

Index(['Span', 'Match', 'Innings', 'Total_Balls', 'Total_Runs', 'Wickets',
       'BBI', 'BBM', 'Average', 'Economy', 'StrikeRate', '5_wkts_in_match',
       '10_wkts_in_match', 'Player_name', 'Country', 'Played_for_ICC'],
      dtype='object')

In [402]:
#rearrenging the column
df = df[['Player_name','Country','Span','Played_for_ICC','Match', 'Innings', 'Total_Balls', 'Total_Runs', 'Wickets',
       'BBI', 'BBM', 'Average', 'Economy', 'StrikeRate', '5_wkts_in_match',
       '10_wkts_in_match']]

In [403]:
#final dataframe after performing 3 question answer
display(df)

Player_name Country  ... 5_wkts_in_match  10_wkts_in_match
0     M Muralitharan       SL  ...              67                22
1           SK Warne      AUS  ...              37                10
2        JM Anderson      ENG  ...              31                 3
3           A Kumble    INDIA  ...              35                 8
4         GD McGrath      AUS  ...              29                 3
5          SCJ Broad      ENG  ...              19                 3
6           CA Walsh       WI  ...              22                 3
7           DW Steyn       SA  ...              26                 5
8        N Kapil Dev    INDIA  ...              23                 2
9       HMRKB Herath       SL  ...              34                 9
10     Sir RJ Hadlee       NZ  ...              36                 9
11          R Ashwin    INDIA  ...              30                 7
12        SM Pollock       SA  ...              16                 1
13   Harbhajan Singh    INDIA  ...              25                 5
14           NM Lyon      AUS  ...              18                 3
15       Wasim Akram      PAK  ...              25                 5
16       CEL Ambrose       WI  ...              22                 3
17           M Ntini       SA  ...              18                 4
18         IT Botham      ENG  ...              27                 4
19       MD Marshall       WI  ...              22                 4
20      Waqar Younis      PAK  ...              22                 5
21        Imran Khan      PAK  ...              23                 6
22        DL Vettori       NZ  ...              20                 3
23         DK Lillee      AUS  ...              23                 7
24        WPUJC Vaas       SL  ...              12                 2
25         AA Donald       SA  ...              20                 3
26        TG Southee       NZ  ...              13                 1
27        RGD Willis      ENG  ...              16                 0
28        MG Johnson      AUS  ...              12                 3
29          I Sharma    INDIA  ...              11                 1
30            Z Khan    INDIA  ...              11                 1
31             B Lee      AUS  ...              10                 0
32          M Morkel       SA  ...               8                 0
33          LR Gibbs       WI  ...              18                 2
34        FS Trueman      ENG  ...              17                 3
35          TA Boult       NZ  ...               9                 1
36      DL Underwood      ENG  ...              17                 6
37         JH Kallis       SA  ...               5                 0
38      CJ McDermott      AUS  ...              14                 2
39          MA Starc      AUS  ...              13                 2
40           BS Bedi    INDIA  ...              14                 1
41    Danish Kaneria      PAK  ...              15                 2
42          J Garner       WI  ...               7                 0
43      JN Gillespie      AUS  ...               8                 0
44          GP Swann      ENG  ...              17                 3
45        JB Statham      ENG  ...               9                 1
46        MA Holding       WI  ...              13                 2
47          R Benaud      AUS  ...              16                 1
48        MJ Hoggard      ENG  ...               7                 1
49       GD McKenzie      AUS  ...              16                 3
50  BS Chandrasekhar    INDIA  ...              16                 2
51         AV Bedser      ENG  ...              15                 5
52         J Srinath    INDIA  ...              10                 1
53       Abdul Qadir      PAK  ...              15                 5
54        Yasir Shah      PAK  ...              16                 3
55         GS Sobers       WI  ...               6                 0
56        AR Caddick      ENG  ...              13                 1
57      

In [404]:
df['Span'].values

array(['1992-2010', '1992-2007', '2003-2022', '1990-2008', '1993-2007',
       '2007-2022', '1984-2001', '2004-2019', '1978-1994', '1999-2018',
       '1973-1990', '2011-2022', '1995-2008', '1998-2015', '2011-2022',
       '1985-2002', '1988-2000', '1998-2009', '1977-1992', '1978-1991',
       '1989-2003', '1971-1992', '1997-2014', '1971-1984', '1994-2009',
       '1992-2002', '2008-2022', '1971-1984', '2007-2015', '2007-2021',
       '2000-2014', '1999-2008', '2006-2018', '1958-1976', '1952-1965',
       '2011-2022', '1966-1982', '1995-2013', '1984-1996', '2011-2022',
       '1966-1979', '2000-2010', '1977-1987', '1996-2006', '2008-2013',
       '1951-1965', '1975-1987', '1952-1964', '2000-2008', '1961-1971',
       '1964-1979', '1946-1955', '1991-2002', '1977-1990', '2014-2021',
       '1954-1974', '1993-2003', '2000-2013', '2012-2021', '2012-2022',
       '2009-2021', '1994-2003', '1946-1960', '2015-2022', '2002-2009',
       '1998-2009', '2011-2020', '2008-2019', '1989-2004', '1925

####Ques 4: Find the no of years played by a player.

In [405]:
#Splitting the Span Column
df_Span = df['Span'].str.split('-', expand = True)
display(df_Span.head(2))


0     1
0  1992  2010
1  1992  2007

In [406]:
#Renaming the Span Column
df_Span = df_Span.rename(columns = {0: 'Debute', 1: 'Retire'})
display(df_Span.head(2))

Debute Retire
0   1992   2010
1   1992   2007

In [407]:
#Concating the df_span & the actual DataFrame
df = pd.concat([df,df_Span], axis = 1)
display(df.head(2))

Player_name Country       Span  ...  10_wkts_in_match Debute  Retire
0  M Muralitharan       SL  1992-2010  ...                22   1992    2010
1        SK Warne      AUS  1992-2007  ...                10   1992    2007

[2 rows x 18 columns]

In [408]:
#droping the Span colummn
df.drop('Span', axis =1, inplace= True)
df.head(2)

Player_name Country  Played_for_ICC  ... 10_wkts_in_match  Debute  Retire
0  M Muralitharan       SL               1  ...               22    1992    2010
1        SK Warne      AUS               0  ...               10    1992    2007

[2 rows x 17 columns]

In [409]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Player_name       80 non-null     object 
 1   Country           80 non-null     object 
 2   Played_for_ICC    80 non-null     int64  
 3   Match             80 non-null     object 
 4   Innings           80 non-null     int64  
 5   Total_Balls       80 non-null     int64  
 6   Total_Runs        80 non-null     int64  
 7   Wickets           80 non-null     int64  
 8   BBI               80 non-null     object 
 9   BBM               80 non-null     object 
 10  Average           80 non-null     float64
 11  Economy           80 non-null     float64
 12  StrikeRate        80 non-null     float64
 13  5_wkts_in_match   80 non-null     int64  
 14  10_wkts_in_match  80 non-null     int64  
 15  Debute            80 non-null     object 
 16  Retire            80 non-null     object 
dtyp

In [410]:
#Convert the string object into int format
df['Retire'] = df['Retire'].astype('int')
df['Debute'] = df['Debute'].astype('int')
df['Match'] = df['Match'].astype('int')

In [411]:
df['years_played'] = df['Retire'] - df['Debute']
df.head(2)

Player_name Country  Played_for_ICC  ...  Debute  Retire  years_played
0  M Muralitharan       SL               1  ...    1992    2010            18
1        SK Warne      AUS               0  ...    1992    2007            15

[2 rows x 18 columns]

In [412]:
#df.columns

In [413]:
#rearrenging the column
df = df[['Player_name', 'Country', 'Debute','Retire','years_played','Played_for_ICC', 'Match', 'Innings',
       'Total_Balls', 'Total_Runs', 'Wickets', 'BBI', 'BBM', 'Average',
       'Economy', 'StrikeRate', '5_wkts_in_match', '10_wkts_in_match']]

In [414]:
df.head(0)

Empty DataFrame
Columns: [Player_name, Country, Debute, Retire, years_played, Played_for_ICC, Match, Innings, Total_Balls, Total_Runs, Wickets, BBI, BBM, Average, Economy, StrikeRate, 5_wkts_in_match, 10_wkts_in_match]
Index: []

#Ques 5: Find the player who have played the most matches in cricket

In [415]:
#finding the 5 top players who have played the most match. 
top_player_based_on_match = df.nlargest(5,'Match')
display(top_player_based_on_match)

Player_name Country  ...  5_wkts_in_match  10_wkts_in_match
2      JM Anderson      ENG  ...               31                 3
37       JH Kallis       SA  ...                5                 0
5        SCJ Broad      ENG  ...               19                 3
1         SK Warne      AUS  ...               37                10
0   M Muralitharan       SL  ...               67                22

[5 rows x 18 columns]

####Ques 5: Replace boolean 'Played_for_ICC' into category 

In [416]:
df['Played_for_ICC'] = df['Played_for_ICC'].replace({0:'No', 1:'Yes'})
df.head(5)

Player_name Country  ...  5_wkts_in_match  10_wkts_in_match
0  M Muralitharan       SL  ...               67                22
1        SK Warne      AUS  ...               37                10
2     JM Anderson      ENG  ...               31                 3
3        A Kumble    INDIA  ...               35                 8
4      GD McGrath      AUS  ...               29                 3

[5 rows x 18 columns]

####Question 6: find the player based on who have given highest run,total wicket taken,total average, strike_rate etc.

In [417]:
def top(x):
  #sorted the values according to query
  df_top = df.sort_values(by= x, ascending= False)
  #reseting the index after sorting
  df_top = df_top.reset_index()
  #removing the redundant column
  df_top = df_top.drop('index', axis= 1)
  #getting the player name by get values
  name_of_the_player = df_top._get_value(0, 'Player_name')
  print('name of the player who have highest ' + x +' is = ', name_of_the_player)


In [418]:
#creating object of top() function
name = top('Total_Runs')
name = top('Average')

name of the player who have highest Total_Runs is =  A Kumble 
name of the player who have highest Average is =  Danish Kaneria 


In [424]:

def plots(key, DATA,y_label):
  
  plt.figure(figsize= (20,12))
  sns.barplot(x = 'Player_name', y = key , data = DATA)
  plt.xlabel('Name of the Player')
  plt.xticks(rotation = 45)
  plt.ylabel(y_label)
  plt.yticks(np.arange(0,110,20))
  plt.title('Top 10 players based on ' + key)
  plt.tight_layout()
  plt.show()
 
  